echopype/preprocess/api.py

```diff
@@ -329,3 +286,433 @@ def remove_noise(ds_Sv, ping_num, range_sample_num, noise_max=None, SNR_threshol
 
 def regrid():
     return 1
+
+
+def get_bin_indices(
+    echo_range: np.ndarray, bins_er: np.ndarray, times: np.ndarray, bins_time: np.ndarray
+) -> Tuple[np.ndarray, np.ndarray]:
+    """
+    Obtains the bin index of ``echo_range`` and ``times`` based
+    on the binning ``bins_er`` and ``bins_time``, respectively.
+
+    Parameters
+    ----------
+    echo_range: np.ndarray
+        2D array of echo range values
+    bins_er: np.ndarray
+        1D array (used by np.digitize) representing the binning required for ``echo_range``
+    times: np.ndarray
+        1D array corresponding to the time values that should be binned
+    bins_time: np.ndarray
+        1D array (used by np.digitize) representing the binning required for ``times``
+
+    Returns
+    -------
+    digitized_echo_range: np.ndarray
+        2D array of bin indices for ``echo_range``
+    bin_time_ind: np.ndarray
+        1D array of bin indices for ``times``
+    """
+
+    # get bin index for each echo range value
+    digitized_echo_range = np.digitize(echo_range, bins_er, right=False)
+
+    # turn datetime into integers, so we can use np.digitize
+    if isinstance(times, dask.array.Array):
+        times_i8 = times.compute().data.view("i8")
+    else:
+        times_i8 = times.view("i8")
+
+    # turn datetime into integers, so we can use np.digitize
+    bins_time_i8 = bins_time.view("i8")
+
+    # get bin index for each time
+    bin_time_ind = np.digitize(times_i8, bins_time_i8, right=False)
+
+    return digitized_echo_range, bin_time_ind
+
+
+def bin_and_mean_echo_range(
+    arr: Union[np.ndarray, dask.array.Array], digitized_echo_range: np.ndarray, n_bin_er: int
+) -> Union[np.ndarray, dask.array.Array]:
+    """
+    Bins and means ``arr`` with respect to the ``echo_range`` bins.
+
+    Parameters
+    ----------
+    arr: np.ndarray or dask.array.Array
+        2D array to bin and mean
+    digitized_echo_range: np.ndarray
+        2D array of bin indices for ``echo_range``
+    n_bin_er: int
+        The number of echo range bins
+
+    Returns
+    -------
+    er_means: np.ndarray or dask.array.Array
+        2D array representing the bin and mean of ``arr`` along ``echo_range``
+    """
+
+    binned_means = []
+    for bin_er in range(1, n_bin_er):
+
+        # Catch a known warning that can occur, which does not impact the results
+        with warnings.catch_warnings():
```

Comment:
@leewujung When we were running the 19 OOI files on data that was in memory, I noticed that a warning was caused by this code block. This warning was spit out by `nanmean` because it was taking the mean of an array filled with NaNs. This still produces the correct output, but the warning is produced. For this reason, I decided to catch specifically this warning and ignore it (all other warnings will **not** be ignored).

Reply:
I see, ok. Sounds good.

Reply:
Thinking about it, this means that somewhere in the 3D array (channel x echo_range x ping_time) there is an NaN "patch" somewhere? This seems a little surprising for OOI data, although I know there are occasional, individual all-NaN pings. For the peace of mind, could you do a quick sanity check to see if there is indeed an all-NaN patch (which results into a bin that would be all NaNs)?

Reply:
@leewujung here is your sanity check that shows some pings are indeed filled with NaNs.
<img width="802" alt="Screen Shot 2022-11-23 at 12 07 27 PM" src="https://user-images.githubusercontent.com/53541061/203637352-1febe4e4-6f0a-4539-b87a-0c2cc9a37bd0.png">

Reply:
Great thanks!